File: ios/Sources/EmojiPicker/Presenter/NativeEmojiPickerPresenter.swift
import UIKit

/// Presents the system emoji keyboard via an injected `EmojiKeyboardPresentationFactory`.
final class NativeEmojiPickerPresenter: EmojiPickerPresenter {

    /// Supplies the current host view controller to present atop; nil if none is available.
    typealias HostViewControllerProvider = () -> UIViewController?

    /// Decides whether a host view controller can currently present UI.
    ///
    /// Isolated from `present` for symmetry with Android's `ActivityAvailabilityChecker` seam and
    /// to keep the check fakeable in tests. UIKit has no throwing-outside-instrumented-tests trap
    /// the way `Activity#isFinishing()` does under Android's mockable jar, so this isn't
    /// load-bearing for crash-avoidance the way Android's is - it exists for seam consistency.
    typealias HostAvailabilityChecker = (UIViewController?) -> Bool

    private static let defaultAvailabilityChecker: HostAvailabilityChecker = { host in
        guard let host = host else { return false }
        return host.isViewLoaded && host.view.window != nil
    }

    private let hostViewControllerProvider: HostViewControllerProvider
    private let factory: EmojiKeyboardPresentationFactory
    private let availabilityChecker: HostAvailabilityChecker

    /// The presentation currently owning the UI/handle, if any. Only the `Presentation` instance
    /// that is still `current` may settle and clear this field; a stale presentation whose async
    /// callback fires after a newer one has taken over must not be able to touch it.
    private var current: Presentation?

    convenience init(hostViewControllerProvider: @escaping HostViewControllerProvider, factory: EmojiKeyboardPresentationFactory) {
        self.init(
            hostViewControllerProvider: hostViewControllerProvider,
            factory: factory,
            availabilityChecker: Self.defaultAvailabilityChecker
        )
    }

    /// Internal: lets tests inject a fake `HostAvailabilityChecker`.
    init(
        hostViewControllerProvider: @escaping HostViewControllerProvider,
        factory: EmojiKeyboardPresentationFactory,
        availabilityChecker: @escaping HostAvailabilityChecker
    ) {
        self.hostViewControllerProvider = hostViewControllerProvider
        self.factory = factory
        self.availabilityChecker = availabilityChecker
    }

    func present(options: EmojiPickerPresentOptions, completion: @escaping (Result<EmojiPickerResult, EmojiPickerError>) -> Void) {
        guard options.presentation == "auto" else {
            // 'web' (and any future non-native presentation) isn't implemented on iOS yet;
            // honestly reject rather than silently showing the native picker instead.
            completion(.failure(EmojiPickerError(code: ErrorCodes.notImplemented)))
            return
        }

        // Capacitor iOS plugin calls are not guaranteed to arrive on the main thread, but
        // UIViewController/UITextField/UIWindow are main-thread-only APIs. Unconditionally hop -
        // unlike Android's Activity-based branch, the provider closure here is always safe to call
        // from any thread, so there's no "no UI to touch" case to special-case around.
        DispatchQueue.main.async { [weak self] in
            self?.presentOnMainThread(
                closeButton: options.closeButton,
                dismissOnBackdropTap: options.dismissOnBackdropTap,
                completion: completion
            )
        }
    }

    private func presentOnMainThread(
        closeButton: EmojiCloseButtonOptions,
        dismissOnBackdropTap: Bool,
        completion: @escaping (Result<EmojiPickerResult, EmojiPickerError>) -> Void
    ) {
        let host = hostViewControllerProvider()
        guard availabilityChecker(host), let host = host else {
            completion(.failure(EmojiPickerError(code: ErrorCodes.nativeUnavailable)))
            return
        }

        // Dismiss the prior presentation's UI, but do NOT clear `current` here: dismissal may
        // complete asynchronously (e.g. dismiss(animated:) completion, or the emoji-mode
        // verification path still in flight). The old Presentation's own settle() - triggered
        // whenever its callback eventually fires - clears `current` itself, and only if it's
        // still the active one, so it can never clobber the new Presentation created below.
        current?.dismissPresentation()

        let presentation = Presentation(completion: completion)
        current = presentation
        presentation.owner = self
        presentation.start(host: host, closeButton: closeButton, dismissOnBackdropTap: dismissOnBackdropTap, factory: factory)
    }

    /// Owns everything about ONE presentation attempt: its handle, its cleanup, and whether it
    /// has already settled. Identity - not a shared field - determines whether a (possibly
    /// late-arriving, async) callback is allowed to mutate `current`.
    private final class Presentation: EmojiKeyboardPresentationListener {
        private let completion: (Result<EmojiPickerResult, EmojiPickerError>) -> Void
        weak var owner: NativeEmojiPickerPresenter?
        private var handle: EmojiKeyboardPresentationHandle?
        private var settled = false
        private var resignActiveObserver: NSObjectProtocol?

        init(completion: @escaping (Result<EmojiPickerResult, EmojiPickerError>) -> Void) {
            self.completion = completion
        }

        func start(
            host: UIViewController,
            closeButton: EmojiCloseButtonOptions,
            dismissOnBackdropTap: Bool,
            factory: EmojiKeyboardPresentationFactory
        ) {
            let createdHandle = factory.create(
                hostViewController: host,
                closeButtonOptions: closeButton,
                dismissOnBackdropTap: dismissOnBackdropTap,
                listener: self
            )

            if settled {
                // Defensive: if create() somehow synchronously settled us (e.g. immediate
                // onUnavailable/onDismissed), don't leak the handle - dismiss it right away.
                createdHandle.dismiss()
                return
            }

            handle = createdHandle

            // Belt-and-suspenders: settle {emoji: nil} if the app backgrounds mid-presentation.
            // There's no exact iOS analog to Android's Lifecycle.ON_DESTROY hook here, but this
            // closes the same class of "never settles" risk if the keyboard/first-responder chain
            // gets torn down by the system while backgrounded.
            resignActiveObserver = NotificationCenter.default.addObserver(
                forName: UIApplication.willResignActiveNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.dismissPresentationAndSettleNil()
            }
        }

        /// Dismisses this presentation's handle, if any. Does not settle synchronously.
        func dismissPresentation() {
            handle?.dismiss()
        }

        private func dismissPresentationAndSettleNil() {
            handle?.dismiss()
            settle(.success(EmojiPickerResult(emoji: nil)))
        }

        // MARK: EmojiKeyboardPresentationListener

        func onEmojiSelected(_ emoji: String) {
            settle(.success(EmojiPickerResult(emoji: emoji)))
        }

        func onDismissed() {
            settle(.success(EmojiPickerResult(emoji: nil)))
        }

        func onUnavailable() {
            settle(.failure(EmojiPickerError(code: ErrorCodes.nativeUnavailable)))
        }

        private func settle(_ result: Result<EmojiPickerResult, EmojiPickerError>) {
            guard !settled else { return }
            settled = true
            closeOut()
            completion(result)
        }

        /// Removes the notification observer and, if still current, clears `current`. Never
        /// touches another Presentation instance's state.
        private func closeOut() {
            if let observer = resignActiveObserver {
                NotificationCenter.default.removeObserver(observer)
            }
            if owner?.current === self {
                owner?.current = nil
            }
        }
    }
}

File: android/src/main/java/app/independo/capacitoremojipicker/presenter/NativeEmojiPickerPresenter.java
package app.independo.capacitoremojipicker.presenter;

import android.app.Activity;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import app.independo.capacitoremojipicker.core.EmojiCloseButtonOptions;
import app.independo.capacitoremojipicker.core.EmojiPickerCallback;
import app.independo.capacitoremojipicker.core.EmojiPickerResult;
import app.independo.capacitoremojipicker.core.ErrorCodes;

/** Presents the native AndroidX emoji picker via an injected {@link EmojiPickerDialogFactory}. */
public class NativeEmojiPickerPresenter implements EmojiPickerPresenter {

    /** Supplies the current hosting Activity; returns null if none is available. */
    public interface ActivityProvider {
        Activity get();
    }

    /**
     * Decides whether an Activity can currently host a dialog.
     *
     * Isolated from {@link #present} so unit tests can inject a fake that never calls real
     * {@code Activity#isFinishing()}/{@code #isDestroyed()} methods, which throw under the
     * Android Gradle Plugin's mockable android.jar outside of Robolectric or instrumented tests.
     */
    interface ActivityAvailabilityChecker {
        boolean isAvailable(Activity activity);
    }

    /** Default checker used in production: mirrors the standard Activity-availability contract. */
    private static final ActivityAvailabilityChecker DEFAULT_AVAILABILITY_CHECKER = (activity) ->
        activity != null && !activity.isFinishing() && !activity.isDestroyed();

    private final ActivityProvider activityProvider;
    private final EmojiPickerDialogFactory dialogFactory;
    private final ActivityAvailabilityChecker availabilityChecker;

    /**
     * The presentation currently owning the dialog/lifecycle-observer, if any. Only the {@link
     * Presentation} instance that is still {@code current} may settle and clear this field; a
     * stale presentation whose async dismissal callback fires after a newer one has taken over
     * must not be able to touch it. See {@link Presentation#settle}.
     */
    private Presentation current;

    public NativeEmojiPickerPresenter(ActivityProvider activityProvider, EmojiPickerDialogFactory dialogFactory) {
        this(activityProvider, dialogFactory, DEFAULT_AVAILABILITY_CHECKER);
    }

    /** Package-private: lets tests inject a fake {@link ActivityAvailabilityChecker}. */
    NativeEmojiPickerPresenter(
        ActivityProvider activityProvider,
        EmojiPickerDialogFactory dialogFactory,
        ActivityAvailabilityChecker availabilityChecker
    ) {
        this.activityProvider = activityProvider;
        this.dialogFactory = dialogFactory;
        this.availabilityChecker = availabilityChecker;
    }

    @Override
    public void present(
        String presentation,
        boolean dismissOnBackdropTap,
        EmojiCloseButtonOptions closeButton,
        String theme,
        EmojiPickerCallback callback
    ) {
        if (!"auto".equals(presentation)) {
            // 'web' (and any future non-native presentation) isn't implemented on Android yet;
            // honestly reject rather than silently showing the native picker instead.
            callback.onError(ErrorCodes.NOT_IMPLEMENTED);
            return;
        }

        Activity activity = activityProvider.get();
        Runnable work = () -> presentOnUiThread(activity, dismissOnBackdropTap, theme, callback);
        // Capacitor invokes @PluginMethod handlers on a background "CapacitorPlugins" thread, but
        // BottomSheetDialog#show() and Lifecycle#addObserver() are UI-thread-only APIs. Hop onto the
        // UI thread whenever we have a real Activity to hop through; a null Activity means there's no
        // UI to touch, so run synchronously and let the availability check below report the error.
        if (activity != null) {
            activity.runOnUiThread(work);
        } else {
            work.run();
        }
    }

    private void presentOnUiThread(Activity activity, boolean dismissOnBackdropTap, String theme, EmojiPickerCallback callback) {
        if (!availabilityChecker.isAvailable(activity)) {
            callback.onError(ErrorCodes.NATIVE_UNAVAILABLE);
            return;
        }

        // Dismiss the prior presentation's dialog, but do NOT clear `current` here: a real
        // BottomSheetDialog#dismiss() delivers its OnDismissListener asynchronously (posted to the
        // message queue), even from the UI thread. The old Presentation's own settle() - triggered
        // whenever that callback eventually fires - will clear `current` itself, and only if it's
        // still the active one, so it can never clobber the new Presentation created below.
        if (current != null) {
            current.dismissDialog();
        }

        Presentation presentation = new Presentation(callback);
        LifecycleOwner lifecycleOwner = (activity instanceof LifecycleOwner) ? (LifecycleOwner) activity : null;
        current = presentation;
        presentation.start(activity, dismissOnBackdropTap, theme, lifecycleOwner);
    }

    /**
     * Owns everything about ONE presentation attempt: its dialog handle, its lifecycle observer,
     * and whether it has already settled. Identity - not a shared field - determines whether a
     * (possibly late-arriving, async) callback is allowed to mutate {@code current}.
     */
    private final class Presentation {

        private final EmojiPickerCallback callback;
        private EmojiPickerDialogFactory.EmojiPickerDialogHandle handle;
        private DefaultLifecycleObserver observer;
        private LifecycleOwner lifecycleOwner;
        private boolean settled;

        Presentation(EmojiPickerCallback callback) {
            this.callback = callback;
        }

        void start(Activity activity, boolean dismissOnBackdropTap, String theme, LifecycleOwner lifecycleOwner) {
            this.lifecycleOwner = lifecycleOwner;

            EmojiPickerDialogFactory.EmojiPickerDialogHandle createdHandle;
            try {
                createdHandle = dialogFactory.create(
                    activity,
                    dismissOnBackdropTap,
                    theme,
                    new EmojiPickerDialogFactory.Listener() {
                        @Override
                        public void onEmojiSelected(String emoji) {
                            settle(new EmojiPickerResult(emoji));
                        }

                        @Override
                        public void onDismissed() {
                            settle(new EmojiPickerResult(null));
                        }
                    }
                );
            } catch (RuntimeException e) {
                // e.g. BadTokenException if the Activity finished in the window between the
                // availability check and dialog creation/show. Settle with NATIVE_UNAVAILABLE
                // instead of letting the exception propagate uncaught out of the UI-thread
                // Runnable, which would crash the app and leave isPresenting stuck forever.
                settleWithError(ErrorCodes.NATIVE_UNAVAILABLE);
                return;
            }

            if (settled) {
                // Defensive: if create() somehow synchronously settled us (e.g. immediate
                // dismissal), don't leak the handle - dismiss it right away.
                try {
                    createdHandle.dismiss();
                } catch (RuntimeException ignored) {
                    // Window may already be gone.
                }
                return;
            }

            this.handle = createdHandle;

            if (lifecycleOwner != null) {
                DefaultLifecycleObserver lifecycleObserver = new DefaultLifecycleObserver() {
                    @Override
                    public void onDestroy(LifecycleOwner owner) {
                        // Dismiss before settling: dismissing a dialog whose window has already
                        // been torn down alongside the Activity can throw (e.g.
                        // IllegalArgumentException: "View not attached to window manager"), which
                        // dismissDialog() guards against.
                        dismissDialog();
                        settle(new EmojiPickerResult(null));
                    }
                };
                this.observer = lifecycleObserver;
                lifecycleOwner.getLifecycle().addObserver(lifecycleObserver);
            }
        }

        /** Dismisses this presentation's dialog handle, if any. Does not settle synchronously. */
        void dismissDialog() {
            if (handle != null) {
                try {
                    handle.dismiss();
                } catch (RuntimeException e) {
                    // Window may already be gone.
                }
            }
        }

        private void settleWithError(String errorCode) {
            if (settled) {
                return;
            }
            settled = true;
            closeOut();
            callback.onError(errorCode);
        }

        private void settle(EmojiPickerResult result) {
            if (settled) {
                return;
            }
            settled = true;
            closeOut();
            callback.onResult(result);
        }

        /** Removes the lifecycle observer and, if still current, clears `current`. Never touches
         * another Presentation instance's state. */
        private void closeOut() {
            if (observer != null && lifecycleOwner != null) {
                lifecycleOwner.getLifecycle().removeObserver(observer);
            }
            if (current == this) {
                current = null;
            }
        }
    }
}
